### PR TITLE
C API Documentation

### PR DESCRIPTION
**Pull request by sslattery**
Doxygen updates for C API.

### CODE REVIEW

**Comment by sslattery** [comment]
What happened with the [cuda build](https://cloud.cees.ornl.gov/jenkins-ci/job/DataTransferKit-GCC54-CUDA90/23/)? Do I need to rebase?

**Comment by Rombur** [comment]
> What happened with the cuda build? Do I need to rebase?

Yes, you do.

File: packages/Interface/src/DTK_C_API.h [hunk 1]
```diff
@@ -385,16 +611,62 @@ typedef void ( *DTK_PolyhedronListSizeFunction )(
...
- *  \param[out] face_orientation Orientation of the faces.
+ *
+ *  \param[out] coordinates Node coordinates. The length of this array is
+ *  local_num_nodes * space_dim. This array is blocked by dimension in
+ *  identical fashion to the NodeList coordinates.
+ *
+ *  \param[out] faces Connectivity list of faces. The length of this array is
+ *  total_face_nodes. This array is defined as rank-1 but represents
+ *  unstructured rank-2 data. It should be sized as (total sum of the number
+ *  of nodes composing each face) or the sum of all elements in the following
```
Comment (dalg24) [inline packages/Interface/src/DTK_C_API.h]: why do you have parentheses here?

File: packages/Interface/src/DTK_C_API.h [hunk 2]
```diff
@@ -50,57 +50,196 @@ extern const char *DTK_version();
...
+ *  following are valid values for the memory space enumeration:
+ *
+ *  DTK_HOST_SPACE: Memory will be allocated on the host in main CPU
+ *  memory. Memory allocated in this space is not directly accessible on the
+ *  GPU. If DTK maps are created for and executed on the GPU or other
+ *  accelerators this memory will be explicitly copied to and from a memory
+ *  space accessible by the GPU for computation.
+ *
+ *  DTK_CUDAUVM_SPACE: Memory will be allocated via CUDA unified virtual
+ *  memory (UVM) for use with an NVIDIA GPU. This memory is automatically
```
[the remark this answers is not in the document]
Comment (sslattery) [reply packages/Interface/src/DTK_C_API.h]: if you sound it out I think it should be "an NVIDIA GPU"

File: packages/Interface/src/DTK_C_API.h [hunk 3]
```diff
@@ -141,17 +286,52 @@ typedef struct _DTK_MapHandle *DTK_MapHandle;
...
+ *  implementations for the target should return zero sizes in allocation
+ *  functions if the user's target application does not exist on the calling
+ *  MPI rank.
+ *
+ *  \param[in] options Options string for building the map. See above for
+ *  details on composing this option string.
+ *
+ *  \return DTK_create returns a handle for the map. This handle must be
+ *  destroyed with DTK_destroyMap() when the lifetime of this map has ended in
+ *  the program.
```
[the remark this answers is not in the document]
Comment (sslattery) [reply packages/Interface/src/DTK_C_API.h]: Thats a good idea - I think that it should. I will make an issue for this.

[the remark this answers is not in the document]
Comment (sslattery) [reply packages/Interface/src/DTK_C_API.h]: see #477

File: packages/Interface/src/DTK_C_API.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (dalg24) [reply packages/Interface/src/DTK_C_API.h]: Fair enough

File: packages/Interface/src/DTK_C_API.h [hunk 4]
```diff
@@ -50,57 +50,202 @@ extern const char *DTK_version();
...
+ *  \param argc Pointer to the number of argument.
+ *  \param argv Pointer to the argument vector.
+ *
+ *  This version of initialize() effectively calls <code>Kokkos::initialize(
+ *  *argc, *argv )</code>.  Pointers to \c argc and \c argv arguments are passed
+ *  in order to match MPI_Init's interface.
+ */
+extern void DTK_initializeCmd( int *argc, char ***argv );
+
+/** \brief Indicates whether DTK has been initialized.
```
Comment (dalg24) [inline packages/Interface/src/DTK_C_API.h]: "Indicate(s)"

File: packages/Interface/src/DTK_C_API.h [hunk 5]
```diff
@@ -50,57 +50,202 @@ extern const char *DTK_version();
...
+ *
+ *  This function may be used to determine whether DTK has been
+ *  initialized. DTK must be initialized before any other library functions
+ *  may be used such as user function callback registration and map creation.
+ */
+extern bool DTK_isInitialized();
+
+/** \brief Finalizes DTK.
```
Comment (dalg24) [inline packages/Interface/src/DTK_C_API.h]: "Finalize(s)"

File: packages/Interface/src/DTK_C_API.h [hunk 6]
```diff
@@ -50,57 +50,202 @@ extern const char *DTK_version();
...
+ * @{
+ */
+
+/** \brief Initialize the DTK execution environment without any arguments.
+ *
+ *  This initializes Kokkos if it has not already been initialized.
+ */
+extern void DTK_initialize();
+
+/** \brief Initialize DTK with command line arguments.
```
[the remark this answers is not in the document]
Comment (sslattery) [reply packages/Interface/src/DTK_C_API.h]: @dalg24 prefers `Initialize`

[the remark this answers is not in the document]
Comment (Rombur) [reply packages/Interface/src/DTK_C_API.h]: But that's the only one without an _s_

File: packages/Interface/src/DTK_C_API.h [hunk 7]
```diff
@@ -355,28 +548,68 @@ typedef void ( *DTK_BoundingVolumeListSizeFunction )(
...
  */
 typedef void ( *DTK_BoundingVolumeListDataFunction )(
     void *user_data, Coordinate *bounding_volumes );
 
 /** \brief Prototype function to get the size parameters for building a
  *  polyhedron list.
  *
- *  Register with a user application using DTK_setUserFunction() by passing
- *  DTK_POLYHEDRON_LIST_SIZE_FUNCTION as the \p type argument.
+ *  A polyehdron list is a collection of arbitrary linear polyhedra defined by
```
[the remark this answers is not in the document]
Comment (sslattery) [reply packages/Interface/src/DTK_C_API.h]: You would give a space dim of 2 and then give no cell information - you just need nodes and faces for polygons. The face orientations at that point are meaningless and would be ignored by the map as one would have to give nodes in an order that gives a positive area when evaluated via the right-hand rule.

File: packages/Interface/src/DTK_C_API.h [hunk 8]
```diff
@@ -385,16 +618,62 @@ typedef void ( *DTK_PolyhedronListSizeFunction )(
...
+ *  \param[out] face_orientation Orientation of the faces. The length of this
+ *  array is total_cell_faces. Orientation of each face composing a cell
+ *  indicating an outward or inward facing normal based on node ordering of
+ *  the face and use of the right-hand rule. This view is defined as rank-1
+ *  but represents unstructured rank-2 data. This view is the same size as the
+ *  cells view and is indexed in an identical matter. If the face for the
+ *  given cell has a node ordering that returns a face normal that points into
+ *  the cell via the right hand rule then a -1 should be input. If the node
+ *  ordering of the face produces a normal that points out from the cell a +1
+ *  should be input.
```
[the remark this answers is not in the document]
Comment (sslattery) [reply packages/Interface/src/DTK_C_API.h]: The face orientations at that point are meaningless and would be ignored by the map as one would have to give nodes in an order that gives a positive area when evaluated via the right-hand rule.

File: packages/Interface/src/DTK_C_API.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (sslattery) [reply packages/Interface/src/DTK_C_API.h]: I have just removed all the `s`'s per @dalg24 review

**Pull request merged**